login-utils/need.8
.\" Copyright (C) 2000  Richard Gooch
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" Richard Gooch may be reached by email at  rgooch@atnf.csiro.au
.\" The postal address is:
.\"   Richard Gooch, c/o ATNF, P. O. Box 76, Epping, N.S.W., 2121, Australia.
.\"
.\"	need.8		Richard Gooch	28-FEB-2000
.\"
.TH NEED 8 "28 Feb 2000" "Util-Linux Package"
.SH NAME
need \- utility to tell simpleinit(8) to start a service
.SH SYNOPSIS
.nf
\fBneed\fP [\fB-r\fP] [\fBservice\fP]
.BR display-services
.fi
.SH DESCRIPTION
The \fBneed\fP programme is a utility that tells \fBsimpleinit\fP(8)
to start a \fIservice\fP (usually a script in \fI/sbin/init.d\fP) and
will wait for the service to become available. If the service is
already available, it will not be started again.

When invoked as \fBdisplay-services\fP it will write the list of
currently available services and the list of failed services to the
standard output.

The \fB-r\fP option is used to tell \fBsimpleinit\fP(8) to "roll back"
(stop) services up to (but not including) \fIservice\fP. If
\fIservice\fP is not specified, all services are stopped. The \fB-r\fP
option thus allows the system to be partially or wholly shut down in
an orderly fashion. The \fBshutdown\fP(8) programme still needs to be
run.

The \fBneed\fP programme is designed to help improve the robustness,
scalability and readability of system boot scripts. It is now possible
to write a modularised set of boot scripts without the complex and
fragile numbered symlink scheme used in SysV-style boot scripts. Each
script can simply declare, using \fBneed\fP(8), what must run before
them.
.SH EXIT CODE
The exit code from \fBneed\fP is 0 if the service was successfully
started, 1 if the service failed badly, and 2 if the service is
unavailable (i.e. disabled in configuration files). These exit codes
reflect the exit codes from the service startup scripts.

The exit code from \fBneed -r\fP is 0 if the service was successfully
stopped, 1 if the service could not be stopped, and 2 if the service
was not available to start with. The service shutdown scripts may only
return 0 (for success) or 1 (for failure).
.SH SIGNALS
\fBneed\fP(8) uses \fBSIGUSR1\fP, \fBSIGUSR2\fP and \fBSIGPOLL\fP for
communication with \fBsimpleinit\fP(8). Don't send these signals to
it.
.SH FILES
.PD 0
.TP 20
.BI /dev/initctl
This is the control FIFO, created by \fBsimpleinit\fP(8), which
\fBneed\fP(8) writes commands to.
.SH SEE ALSO
.BR simpleinit (8),
.BR init (8)
.PP
A more complete discussion of the new boot script system, based on
\fBneed\fP(8), is available from:
http://www.atnf.csiro.au/~rgooch/linux/boot-scripts/
.SH AUTHOR
Richard Gooch (rgooch@atnf.csiro.au)
.SH AVAILABILITY
The Util-Linux package is available from:
ftp://ftp.??.kernel.org/pub/linux/utils/util-linux/


login-utils/need.c
/*  need.c

    Source file for  need  (init(8) dependency tool).

    Copyright (C) 2000  Richard Gooch

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    Richard Gooch may be reached by email at  rgooch@atnf.csiro.au
    The postal address is:
      Richard Gooch, c/o ATNF, P. O. Box 76, Epping, N.S.W., 2121, Australia.
*/

/*
    This tool will request init(8) to start a service and will wait for that
    service to be available. If the service is already available, init(8) will
    not start it again.
    This tool may also be used to inspect the list of currently available
    services.


    Written by      Richard Gooch   28-FEB-2000

    Last updated by Richard Gooch   28-FEB-2000


*/
#include <unistd.h>
#include <stdio.h>
#include <limits.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <signal.h>
#include <string.h>
#include "simpleinit.h"


static void signal_handler (int sig);


static int caught_signal = 0;


int main (int argc, char **argv)
{
    int fd, nbytes;
    struct sigaction sa;
    sigset_t ss;
    char *ptr;
    long buffer[COMMAND_SIZE / sizeof (long)];
    struct command_struct *command = (struct command_struct *) buffer;

    sigemptyset (&ss);
    sigaddset (&ss, SIG_PRESENT);
    sigaddset (&ss, SIG_NOT_PRESENT);
    sigaddset (&ss, SIG_FAILED);
    sigprocmask (SIG_BLOCK, &ss, NULL);
    sigemptyset (&sa.sa_mask);
    sa.sa_flags = 0;
    sa.sa_handler = signal_handler;
    sigaction (SIG_PRESENT, &sa, NULL);
    sigaction (SIG_NOT_PRESENT, &sa, NULL);
    sigaction (SIG_FAILED, &sa, NULL);
    command->pid = getpid ();
    if ( ( ptr = strrchr (argv[0], '/') ) == NULL ) ptr = argv[0];
    else ++ptr;
    if (strcmp (ptr, "display-services") == 0)
    {
	command->command = COMMAND_DUMP_LIST;
	if (tmpnam (command->name) == NULL)
	{
	    fprintf (stderr, "Unable to create a unique filename\t%s\n",
		     ERRSTRING);
	    exit (1);
	}
	if (mkfifo (command->name, S_IRUSR) != 0)
	{
	    fprintf (stderr, "Unable to create FIFO: \"%s\"\t%s\n",
		     command->name, ERRSTRING);
	    exit (1);
	}
    }
    else
    {
	if ( (argc > 1) && (strcmp (argv[1], "-r") == 0) )
	{
	    command->command = COMMAND_ROLLBACK;
	    --argc;
	    ++argv;
	}
	else command->command = COMMAND_NEED;
	if (argc == 2) strcpy (command->name, argv[1]);
	else
	{
	    if (command->command == COMMAND_ROLLBACK) command->name[0] = '\0';
	    else
	    {
		fprintf (stderr, "Usage:\tneed programme\n");
		exit (1);
	    }
	}
    }
    if ( ( fd = open ("/dev/initctl", O_WRONLY, 0) ) < 0 )
    {
	fprintf (stderr, "Error opening\t%s\n", ERRSTRING);
	exit (1);
    }
    if (write (fd, buffer, COMMAND_SIZE) < COMMAND_SIZE)
    {
	fprintf (stderr, "Error writing\t%s\n", ERRSTRING);
	exit (1);
    }
    close (fd);
    if (command->command != COMMAND_DUMP_LIST)
    {
	sigemptyset (&ss);
	while (caught_signal == 0) sigsuspend (&ss);
	switch (caught_signal)
	{
	  case SIG_PRESENT:
	    return 0;
	  case SIG_NOT_PRESENT:
	    return 2;
	  case SIG_FAILED:
	    return 1;
	}
	return 3;
    }
    /*  Read back the data and display it  */
    if ( ( fd = open (command->name, O_RDONLY, 0) ) < 0 )
    {
	fprintf (stderr, "Error opening:\"%s\"\t%s\n",
		 command->name, ERRSTRING);
	exit (1);
    }
    unlink (command->name);
    fflush (stdout);
    while ( ( nbytes = read (fd, buffer, COMMAND_SIZE) ) > 0 )
	write (1, buffer, nbytes);
    close (fd);
    return (0);
}   /*  End Function main  */

static void signal_handler (int sig)
{
    caught_signal = sig;
}   /*  End Function signal_handler  */
